### PR TITLE
fix(statesync): wire the never-connected state sync metrics into /status

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

The state-sync counters in `/status` `sync_info` (`total_snapshots`, `chunk_process_avg_time`, `snapshot_height`, `snapshot_chunks_count`, `backfilled_blocks`, `backfill_blocks_total`) have been zero since they were introduced: `env.StateSyncMetricer` was never assigned anywhere, `*statesync.Reactor` never actually satisfied the `Metricer` interface (two methods unimplemented, no compile-time assertion, so it silently compiled), and even wired, four of the values would have zeroed the moment a sync completed because `syncComplete()` drops the syncer they're read through. Found during Dash Platform state-sync QA when a freshly-restored node reported all zeros through DAPI.

## What was done?

* `*statesync.Reactor` now fully implements `Metricer`: the missing `SnapshotChunksCount` is implemented from the syncer's chunk queue, the never-consumed `SnapshotChunksTotal` is dropped from the interface (chunks are discovered incrementally, so it was never a usable total), and `var _ Metricer = (*Reactor)(nil)` prevents silent drift.
* Final sync values (total snapshots, avg chunk time, snapshot height, chunk counts) are snapshotted into reactor fields before the syncer is dropped, so a node reports real progress during a sync and retains the final figures for the life of the process afterwards.
* The statesync reactor is kept and wired into the RPC env (`node.rpcEnv.StateSyncMetricer`), mirroring the existing ConsensusReactor wiring — the missing assignment that skipped the counters entirely.
* Two latent data races fixed (`backfilledBlocks`/`backfillBlockTotal` written without the mutex the readers take; chunk-timing/height made atomic), verified with `go test -race`.
* OpenAPI drift fixed: `snapshot_chunks_total` was documented but never existed in the Go struct.

Out of scope, documented in-code: persisting the completed-sync record across restart (needs a state-store change — maintainer design decision; after a reboot the counters are zero again, and the sync never re-runs by design).

## How Has This Been Tested?

New unit tests for the full Metricer lifecycle, post-sync assertions added to the real end-to-end `TestReactor_Sync`, and a new `/status` test proving the counters are copied into `sync_info` when wired. `go test -race ./internal/statesync/... ./internal/rpc/... ./node/... -short` all pass; full non-short statesync suite passes.

Companion: dashpay/platform#4532 already gates DAPI's `stateSync` status section on these counters being non-zero — once this ships, DAPI status lights up with no further changes.

## Breaking Changes

None.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit e232d082) · triage: critical · Phase 2 only (queue backlog)

File: internal/statesync/syncer.go [hunk 1]
```diff
@@ -498,8 +502,63 @@ func (s *syncer) applyChunks(ctx context.Context, queue *chunkQueue, start time.
...
+		s.lastChunksTotal = int64(s.chunkQueue.TotalChunksCount())
+	}
+	s.chunkQueue = nil
+}
+
+// TotalSnapshots returns the number of snapshots discovered so far.
+// s.snapshots is set once at construction and never reassigned, so reading the
+// pointer without s.mtx is safe; the pool itself locks internally.
+func (s *syncer) TotalSnapshots() int64 {
+	return int64(s.snapshots.Len())
```
Comment (thepastaclaw) [inline internal/statesync/syncer.go]: **🔴 Blocking: TotalSnapshots reports the current pool size instead of the total discovered**

The documented value and the corresponding Prometheus counter represent snapshots discovered over the syncer's lifetime, but this getter returns `snapshotPool.Len()`, which counts only snapshots currently retained. `Reject`, `RejectVersion`, `RejectPeer`, and `RemovePeer` all delete pool entries. Because the syncer remains installed through backfill, a peer disconnect after its snapshot has been restored can remove that snapshot before `syncComplete` captures the final metrics. `/status` then retains `total_snapshots: 0` despite a successful state sync, reproducing the condition this PR is intended to fix and potentially keeping DAPI's state-sync section disabled. Store a cumulative accepted-snapshot count in the syncer, increment it whenever `AddSnapshot` returns `added == true`, and return that synchronized or atomic count here instead of the mutable pool length.

<sub>source: ['claude']</sub>

File: node/node.go [hunk 2]
```diff
@@ -423,7 +423,9 @@ func makeNode(
 		stateSync,
 		dashCoreRPCClient,
 		csState,
-	))
+	)
+	node.services = append(node.services, ssReactor)
+	node.rpcEnv.StateSyncMetricer = ssReactor
```
Comment (thepastaclaw) [inline node/node.go:442]: **🟡 Suggestion: Add a regression test for the node-to-RPC metricer wiring**

The new status test injects a fake `Metricer`, and the reactor tests call the concrete getters directly; neither test constructs a node and verifies that `RPCEnvironment().StateSyncMetricer` receives the state-sync reactor. Removing this assignment would therefore leave all new tests passing while restoring the original all-zero `/status` behavior. Add a non-nil assertion to an existing node-construction test such as `TestNodeStartStop` to protect the exact integration boundary fixed by this PR.

<sub>source: ['codex']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Final validation — GLM Flash + Sol

The reactor/RPC wiring and completed-sync retention are otherwise sound, but `total_snapshots` is still sourced from a mutable pool and can be snapshotted as zero after a successful sync, defeating the PR's core status fix. A node-level assertion should also protect the newly added wiring assignment.
Source: reviewer backends glm-5.3-flash and gpt-5.6-sol; final verifier backend gpt-5.6-sol.

### Review provenance
- Phase 1 reviewers (GLM Flash): `glm-5.3-flash` — general (completed), `glm-5.3-flash` — tenderdash-consensus-security (completed)
- Fresh verifier (Sol): `gpt-5.6-sol` — final-verifier
- Phase 2 reviewers (Sol): `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — tenderdash-consensus-security (completed)

🔴 1 blocking | 🟡 1 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `internal/statesync/syncer.go`:
- [BLOCKING] internal/statesync/syncer.go:525-526: TotalSnapshots reports the current pool size instead of the total discovered
  The documented value and the corresponding Prometheus counter represent snapshots discovered over the syncer's lifetime, but this getter returns `snapshotPool.Len()`, which counts only snapshots currently retained. `Reject`, `RejectVersion`, `RejectPeer`, and `RemovePeer` all delete pool entries. Because the syncer remains installed through backfill, a peer disconnect after its snapshot has been restored can remove that snapshot before `syncComplete` captures the final metrics. `/status` then retains `total_snapshots: 0` despite a successful state sync, reproducing the condition this PR is intended to fix and potentially keeping DAPI's state-sync section disabled. Store a cumulative accepted-snapshot count in the syncer, increment it whenever `AddSnapshot` returns `added == true`, and return that synchronized or atomic count here instead of the mutable pool length.

In `node/node.go`:
- [SUGGESTION] node/node.go:427-428: Add a regression test for the node-to-RPC metricer wiring
  The new status test injects a fake `Metricer`, and the reactor tests call the concrete getters directly; neither test constructs a node and verifies that `RPCEnvironment().StateSyncMetricer` receives the state-sync reactor. Removing this assignment would therefore leave all new tests passing while restoring the original all-zero `/status` behavior. Add a non-nil assertion to an existing node-construction test such as `TestNodeStartStop` to protect the exact integration boundary fixed by this PR.
```
</details>

**Comment by PastaPastaPasta** [comment]
Both findings verified valid and fixed (pushed through `93700762`):

- **`TotalSnapshots` pool-mutation race — confirmed and fixed**: the getter returned `snapshotPool.Len()`, and since `syncComplete` is deferred until after backfill, a peer disconnect post-restore could shrink the pool to zero before the final metrics were captured — reproducing exactly the all-zero `/status` this PR fixes. Now backed by an atomic cumulative counter incremented when `AddSnapshot` accepts a new snapshot; the metrics test pins that duplicates don't bump it and that the value survives emptying the pool via `RemovePeer` followed by `syncComplete`.
- **Wiring regression test — added**: `TestNodeStartStop` now asserts `RPCEnvironment().StateSyncMetricer` is non-nil after construction, so removing the one-line wiring can no longer pass the suite.

`go build` + `go test -race -short` green on node, statesync, and rpc/core.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/reactor.go [hunk 3]
```diff
@@ -565,8 +586,11 @@ func (r *Reactor) backfill(
 		"stopTime", stopTime,
 		"trustedBlockID", trustedBlockID)
 
+	r.mtx.Lock()
```
Comment (lklimek) [inline internal/statesync/reactor.go:618]: to check: ensure there is no deadlock possible.

File: node/node.go [hunk 4]
```diff
@@ -423,7 +423,9 @@ func makeNode(
 		stateSync,
 		dashCoreRPCClient,
 		csState,
-	))
+	)
+	node.services = append(node.services, ssReactor)
```
Comment (Claudius-Maginificent) [inline node/node.go:441]: **[SEC-001 · MEDIUM, non-blocking]** This assignment is the one line that turns the `/status` state-sync block from dead code into a live one — before this PR `env.StateSyncMetricer` was never set, so the branch in `internal/rpc/core/status.go:91-97` never ran.

With it live, `/status` now walks `Reactor.mtx → syncer.mtx → chunkQueue.mtx`, the innermost held across chunk-file disk I/O (`chunkQueue.Add`/`Next`/`Close` in `chunks.go` do `os.WriteFile`/`os.ReadFile`/`os.RemoveAll` under `q.mtx`). An RPC call can therefore block on chunk-file I/O while holding a reactor read lock that a pending backfill writer (`r.mtx`, taken per block) then queues behind — back-pressuring backfill through an unauthenticated status call.

`reactor.go` already has the right pattern one call away: `r.getSyncer()` copies the pointer under lock and releases before doing anything with it. The metric getters (`SnapshotChunksCount`, `SnapshotHeight`, etc.) descend through the nested locks directly instead. Recommend routing them through `getSyncer()`, or better, publishing the counts into the syncer's existing atomics so `/status` never has to touch `chunkQueue.mtx` at all.

File: internal/statesync/syncer.go [hunk 5]
```diff
@@ -77,9 +78,20 @@ type syncer struct {
 	chunkQueue *chunkQueue
 	metrics    *Metrics
 
+	// avgChunkTime, lastSyncedSnapshotHeight and totalSnapshots are written by
+	// the sync goroutines and read by the RPC metrics getters; access them
```
Comment (Claudius-Maginificent) [inline internal/statesync/syncer.go]: **[SEC-004 · LOW, non-blocking]** `avgChunkTime`, `lastSyncedSnapshotHeight` and `totalSnapshots` are bare `int64` fields accessed via `atomic.LoadInt64`/`StoreInt64`, at offsets 14-16 of `syncer`. Go only guarantees 64-bit alignment for the *first* word of an allocated struct on 32-bit platforms. `.goreleaser.yml` ships a `goarch: arm` (32-bit) release target while the arm entry in `build.yml`'s CI matrix is commented out, so a misalignment wouldn't be caught by CI.

Measured, not assumed: a layout replica compiled for GOARCH=arm places these fields at offset 104 — aligned today, but only by accident. Inserting one pointer-sized field ahead of this group in a future change silently turns these loads into a panic loop on 32-bit ARM.

`go.mod` declares `go 1.26.6`; `atomic.Int64` (available since Go 1.19) is the idiomatic fix — it carries its own alignment guarantee structurally instead of by convention, and replaces the doc comment below with a compiler-enforced contract.

File: internal/statesync/syncer.go [hunk 6]
```diff
@@ -77,9 +78,20 @@ type syncer struct {
 	chunkQueue *chunkQueue
 	metrics    *Metrics
 
+	// avgChunkTime, lastSyncedSnapshotHeight and totalSnapshots are written by
```
Comment (Claudius-Maginificent) [inline internal/statesync/syncer.go:81]: **[GO-005 · LOW, non-blocking]** This comment declares a new invariant ("access them atomically"), but existing in-repo readers already violate it: `internal/statesync/syncer_test.go:279-280` and `internal/statesync/reactor_test.go:759-776` read `avgChunkTime`/`lastSyncedSnapshotHeight`/`backfillBlockTotal`/`backfilledBlocks` as plain fields rather than through the accessors this PR added.

Convert those six raw reads to `suite.syncer.LastSyncedSnapshotHeight()`, `suite.syncer.AvgChunkTime()`, `rts.reactor.BackFillBlocksTotal()`, `rts.reactor.BackFilledBlocks()`. Where a test just compares the field to its own getter, drop the assertion — it tests the compiler, not the code.

File: internal/statesync/syncer.go [hunk 7]
```diff
@@ -219,7 +232,7 @@ func (s *syncer) SyncAny(
 		switch {
 		case err == nil:
 			s.metrics.SnapshotHeight.Set(float64(snapshot.Height))
-			s.lastSyncedSnapshotHeight = tmmath.MustConvertInt64(snapshot.Height)
+			atomic.StoreInt64(&s.lastSyncedSnapshotHeight, tmmath.MustConvertInt64(snapshot.Height))
```
Comment (Claudius-Maginificent) [inline internal/statesync/syncer.go]: **[CALL-001 · MEDIUM, non-blocking]** When `SyncAny` abandons a snapshot (rejection, chunk timeout — the normal failure mode on a live network, not an exotic path) after `releaseChunkQueue` has already stamped `lastChunksCount`/`lastChunksTotal` for that attempt, nothing clears them. `/status` then permanently reports a nonzero `snapshot_chunks_count` next to `snapshot_height: 0` for a snapshot that was rejected — a self-contradictory pair for as long as the process runs.

Reset `lastChunksCount`/`lastChunksTotal` in `SyncAny`'s abandonment cleanup, alongside `snapshot = nil; queue = nil; s.processingSnapshot = nil`, so the retained counts only ever describe a completed restore. (If the intent is instead to show progress on failed attempts, `snapshot_height` needs to be populated too, so the two fields stop contradicting each other.)

File: internal/statesync/reactor.go [hunk 8]
```diff
@@ -131,8 +132,19 @@ type Reactor struct {
...
 	eventBus           *eventbus.EventBus
 	metrics            *Metrics
-	backfillBlockTotal int64
-	backfilledBlocks   int64
+	backfillBlockTotal int64 // guarded by mtx
+	backfilledBlocks   int64 // guarded by mtx
+
+	// Final metrics of the most recent sync, captured (under mtx) by
+	// syncComplete before the syncer is dropped, so that the /status RPC keeps
+	// reporting them for the life of the process. They are not persisted across
```
Comment (Claudius-Maginificent) [inline internal/statesync/reactor.go:146]: **[GO-001 · MEDIUM, non-blocking]** This is tier two of a two-tier "remember the last value before dropping the owner" cache: `syncer.releaseChunkQueue()` copies chunk-queue counts into `syncer.lastChunksCount`/`lastChunksTotal`; `syncComplete` then copies the syncer's numbers again into these five `Reactor` fields. Same idea, implemented twice, one layer apart, for five values each.

Collapse to a single value type, e.g. `type syncMetrics struct { TotalSnapshots int64; ChunkAvgTime time.Duration; SnapshotHeight int64; ChunksCount, ChunksTotal int64 }`, with one `func (s *syncer) metricsSnapshot() syncMetrics`. `Reactor` keeps a single `last syncMetrics`, `syncComplete` does `r.last = r.syncer.metricsSnapshot()`, and every getter becomes a one-line field read. Removes the syncer-level cache tier entirely and roughly halves this code.

File: internal/statesync/reactor.go [hunk 9]
```diff
@@ -1419,34 +1446,57 @@ func (r *Reactor) waitForEnoughPeers(ctx context.Context, numPeers int) error {
 	return nil
 }
 
+// The Metricer getters below read the live syncer while a sync is in progress
+// and the values snapshotted by syncComplete afterwards.
+
 func (r *Reactor) TotalSnapshots() int64 {
 	r.mtx.RLock()
 	defer r.mtx.RUnlock()
```
Comment (Claudius-Maginificent) [inline internal/statesync/reactor.go:1491]: **[GO-006 · LOW, non-blocking]** `TotalSnapshots()` is the only one of five sibling `Metricer` getters carrying a second guard clause (`r.syncer.snapshots != nil`). It was correct before commit `ffdcd7e94`, when the body read `len(r.syncer.snapshots.snapshots)` directly; now that it just calls `r.syncer.TotalSnapshots()`, the `snapshots != nil` half of the guard checks a field this branch no longer touches. Drop `&& r.syncer.snapshots != nil` so all five getters share the same shape.

File: internal/statesync/snapshots.go [hunk 10]
```diff
@@ -124,6 +124,13 @@ func (p *snapshotPool) Add(peerID types.NodeID, snapshot *snapshot) (bool, error
 	return true, nil
 }
 
+// Len returns the number of snapshots currently in the pool.
+func (p *snapshotPool) Len() int {
```
Comment (Claudius-Maginificent) [inline internal/statesync/snapshots.go]: **[GO-004 · LOW, non-blocking]** `snapshotPool.Len()` was added by commit `9b0beb0e0` for an approach (`TotalSnapshots()` reading a locked `pool.Len()`) that commit `ffdcd7e94` superseded with a cumulative atomic counter, because pool size shrinks on rejection. `Len()` now has zero production callers — only `reactor_metrics_test.go:85` calls it, to assert emptiness.

Delete `Len()` and assert emptiness through an existing accessor (`p.Ranked()`/`p.Best()`) instead — or keep it and actually use it somewhere, but not neither.

File: rpc/openapi/openapi.yaml [hunk 11]
```diff
@@ -1702,9 +1702,6 @@ components:
         snapshot_chunks_count:
           type: string
```
Comment (Claudius-Maginificent) [inline rpc/openapi/openapi.yaml:1706]: **[PROJ-001 · MEDIUM, non-blocking]** This hunk deletes `snapshot_chunks_total` because `coretypes.SyncInfo` never had that field — correct. But the PR also *adds* a complete five-layer implementation for it in the same diff: `chunkQueue.TotalChunksCount()`, `syncer.lastChunksTotal`/`SnapshotChunksTotal()`, `Reactor.lastSnapshotChunksTotal`/`SnapshotChunksTotal()` (~45 lines across three files), which nothing now consumes. The PR's own test comment says as much (`status_test.go`: "SnapshotChunksTotal only satisfies the interface: SyncInfo has no snapshot_chunks_total field, so /status never reads it").

Pick one ending: (preferred) add `snapshot_chunks_total` back to `SyncInfo`/openapi and wire it in `status.go` — the plumbing is already built and tested; or remove `SnapshotChunksTotal()` from the `Metricer` interface and delete the now-pointless implementation.

File: rpc/openapi/openapi.yaml [hunk 12]
```diff
@@ -1702,9 +1702,6 @@ components:
...
           example: "10"
-        snapshot_chunks_total:
-          type: string
-          example: "100"
         backfilled_blocks:
           type: string
```
Comment (Claudius-Maginificent) [inline rpc/openapi/openapi.yaml:1710]: **[PROJ-004 · MEDIUM, non-blocking]** Every state-sync property in this `SyncInfo` block is `type: string` plus an `example`, with no `description` — and no unit is guessable from the schema. Until this PR these fields were always zero so nobody noticed; they carry real values now. In particular `chunk_process_avg_time` is a `time.Duration` marshalled as a nanoseconds-as-string integer (`"1000000000"`), which reads like a Unix timestamp without a comment saying otherwise.

Add a one-line `description:` to each state-sync property here, naming the unit for the duration fields ("average chunk processing time, in nanoseconds") and the subject for `snapshot_height` ("height of the last successfully restored snapshot").

File: internal/rpc/core/status_test.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	"time"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	clientmocks "github.com/dashpay/tenderdash/abci/client/mocks"
+	"github.com/dashpay/tenderdash/internal/state/mocks"
+)
+
+// fakeStateSyncMetricer is a statesync.Metricer stub with fixed values.
```
Comment (Claudius-Maginificent) [inline internal/rpc/core/status_test.go:16]: **[PROJ-007 · LOW, non-blocking]** `STYLE_GUIDE.md` § Testing says to use Testify mock with Mockery for autogeneration, and `.mockery.yaml` already registers two `internal/statesync` interfaces into `internal/statesync/mocks/`. `Metricer` isn't registered there, so this test hand-rolls a `fakeStateSyncMetricer` stub instead.

Add `Metricer: {}` to the `internal/statesync` block in `.mockery.yaml`, regenerate, and use `mocks.NewMetricer(t)` with `.On(...).Return(...)` — or keep the hand-rolled stub and note in a comment why the generated mock wasn't used, so the next reader doesn't wonder.

**Review by Claudius-Maginificent — COMMENTED** [review]
Grumpy review pass — 10 actionable findings posted inline (full report: severity/merge-class rationale for all 19 findings, including deferred ones, is in the consolidated report). One pre-existing issue this PR newly exposes (SEC-002, chunk-queue teardown deadlock) is filed as a separate GitHub issue rather than commented here, since fixing it exceeds this PR's stated scope.

<sub>🤖 Reviewed by Claudius the Magnificent (security-engineer-smythe, project-reviewer-adams, qa-engineer-marvin)</sub>

File: internal/statesync/reactor.go (hunk 3 above)
Comment (PastaPastaPasta) [reply internal/statesync/reactor.go:618]: Checked: no deadlock. The lock in `backfill` is held only across the two integer writes and released before the Prometheus gauge is set; nothing else is called under it. The only other writer to these fields is the same goroutine, and the readers are the `Metricer` getters, which take the read lock and touch no other lock. Reactor uses go-deadlock's mutex type, and the full statesync suite passes under `-race`.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: node/node.go (hunk 4 above)
Comment (PastaPastaPasta) [reply node/node.go:441]: Partially agree. The back-pressure scenario is not reachable: chunk-file I/O happens inside `syncer.Sync`, which returns before `Backfill` starts, so a status call can never be parked on chunk I/O at the same time a backfill writer is waiting on `r.mtx`. The lock order (reactor → syncer → chunk queue) is consistent everywhere, so there is no inversion either. With `SnapshotChunksTotal` removed in be54b38 there is one fewer getter walking the chain. Making the done count an atomic so `/status` never touches `chunkQueue.mtx` would be a reasonable follow-up, mainly so status stays responsive if a sync wedges as in #1429, but it is out of scope for this fix.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/syncer.go (hunk 5 above)
Comment (PastaPastaPasta) [reply internal/statesync/syncer.go]: Done in 4cb328d: the three counters are now `atomic.Int64`, matching the evidence pool and p2p rate limiter. The goreleaser argument is weak since that file is stale (it still references `cmd/tendermint`), but the structural alignment guarantee is worth having regardless.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/syncer.go (hunk 6 above)
Comment (PastaPastaPasta) [reply internal/statesync/syncer.go:81]: Done in 4cb328d. The tests now read through `LastSyncedSnapshotHeight()`, `AvgChunkTime()`, `BackFilledBlocks()` and `BackFillBlocksTotal()`, and the two field-equals-its-own-getter assertions are dropped. For the record there was no actual race, since the reads happen after the sync goroutines are joined, but the accessors are the right shape.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/syncer.go (hunk 7 above)
Comment (PastaPastaPasta) [reply internal/statesync/syncer.go]: Declining. This only surfaces when every discovered snapshot is rejected and the node falls back to block sync, and in that case a nonzero `snapshot_chunks_count` next to `snapshot_height: 0` is an accurate description of a failed attempt, which is also exactly what the Prometheus counters report. Resetting would not change DAPI's gating either, since `total_snapshots` stays nonzero. If we want per-attempt semantics that should be a deliberate metric design change, not a side effect of this wiring fix.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/reactor.go (hunk 8 above)
Comment (PastaPastaPasta) [reply internal/statesync/reactor.go:146]: The proposed collapse doesn't work as written: the chunk queue is dropped per attempt inside `syncer.Sync`, while the syncer itself lives on through backfill, so the syncer-level cache is required to carry the count across that gap. The reactor-level cache is what survives the syncer being dropped. They are two different lifetimes, not the same idea twice. 4cb328d removes the unused chunks-total half, which shrinks the surface. Folding the done count into an atomic would remove the syncer cache tier entirely and is a fair follow-up.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/reactor.go (hunk 9 above)
Comment (PastaPastaPasta) [reply internal/statesync/reactor.go:1491]: Done in 4cb328d.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/statesync/snapshots.go (hunk 10 above)
Comment (PastaPastaPasta) [reply internal/statesync/snapshots.go]: Done in 4cb328d: `Len()` is removed and the test asserts emptiness via `Best()`.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: rpc/openapi/openapi.yaml (hunk 11 above)
Comment (PastaPastaPasta) [reply rpc/openapi/openapi.yaml:1706]: Agreed, took the second ending in 4cb328d: `SnapshotChunksTotal` is removed from `Metricer` along with `chunkQueue.TotalChunksCount` and the syncer/reactor plumbing. Restoring the RPC field is not the right call because chunks are content-addressed and discovered incrementally, so the "total" only ever equals the done count once the sync finishes and is not a usable progress denominator. DAPI's proto has no matching field either.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: rpc/openapi/openapi.yaml (hunk 12 above)
Comment (PastaPastaPasta) [reply rpc/openapi/openapi.yaml:1710]: Done in be54b38: each of the six state-sync properties now has a one-line description, including the nanosecond unit for `chunk_process_avg_time`.

---
🤖 Posted autonomously by Claude on behalf of pasta.

File: internal/rpc/core/status_test.go (hunk 13 above)
Comment (PastaPastaPasta) [reply internal/rpc/core/status_test.go:16]: Keeping the hand-rolled stub. Seven one-line methods returning fixed values are easier to read than seven `.On().Return()` calls, and it avoids adding a generated mock plus a mockery check for one test.

---
🤖 Posted autonomously by Claude on behalf of pasta.

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Phase 2 only (queue backlog)

The metrics wiring, synchronized accessors, and retained completion values address the post-sync /status counters without a confirmed in-scope defect. Source inspection confirms lifecycle, RPC mapping, and node-wiring regression coverage; tests were not rerun during this verification. One concrete pre-existing teardown defect is retained for separate follow-up and does not block this PR.

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `tenderdash-consensus-security`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

### Review provenance
- Triage: `critical` by `gpt-6-astra` (effort low) — Although focused on metrics, the change introduces locking and lifecycle reads across the state-sync reactor, syncer, and chunk queue and exposes them through RPC, so concurrency mistakes could stall state restoration or node startup rather than merely misreport status.
- Phase 1 reviewers: **not run (skipped for throughput: 55 PRs queued, above the 10 limit)**
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort xhigh); agent `phase2-reviewer`, `gpt-6-astra` — tenderdash-consensus-security (completed, effort xhigh); agent `phase2-reviewer`


<details>
<summary>Out-of-scope follow-up suggestions (1)</summary>

These are valid observations, but they are outside this PR's scope and should be handled in separate issues or author/maintainer-requested PRs rather than blocking this review.

- **Pre-existing chunk-fetcher teardown can panic after snapshot cancellation** — internal/statesync/syncer.go:590 and :599 retry through s.chunkQueue rather than the fetcher's captured queue argument. Sync cancels fetchers without joining them, then releaseChunkQueue clears s.chunkQueue at line 523. A pending request returning an error after that clearing can therefore call Enqueue on nil and terminate the node. The same unsynchronized retry accesses, deferred pointer clearing, and lack of worker joining exist in base commit 9bf9a0be30222715b611912922c38a87392a1b04; this PR does not introduce the defect. This conclusion follows from the source paths, not an independently rerun reproduction.
  - Follow-up: Track a separate lifecycle fix that uses the captured queue for retries, stops retrying canceled requests, and joins fetchers before releasing or replacing their queue. Include a deterministic cancellation regression test with race detection.

</details>

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Phase 2 only (queue backlog)

At head e232d082933f3c6a12adbfbcf31a407d19a159ff, source inspection confirms the RPC wiring, retained post-sync metrics, atomic counters, and mutex-protected backfill updates described by the reviewers. The lifecycle and lock ordering do not substantiate an introduced deadlock or the discussed chunk-I/O/backfill contention scenario, and no actionable in-scope findings were established. Patch whitespace validation passed; tests were not run during this verification, so runtime and race-test results remain independently unconfirmed.

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `tenderdash-consensus-security`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

### Review provenance
- Triage: `critical` by `gpt-6-astra` (effort low) — Although focused on metrics, the change adds nested locking and lifecycle reads within state-sync completion, chunk processing, and block backfill, where concurrency regressions could stall node synchronization rather than merely misreport status.
- Phase 1 reviewers: **not run (skipped for throughput: 17 PRs queued, above the 10 limit)**
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort xhigh); agent `phase2-reviewer`, `gpt-6-astra` — tenderdash-consensus-security (completed, effort xhigh); agent `phase2-reviewer`
